server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -1853,12 +1853,34 @@ public void close(String reason, boolean flushEngine) throws IOException {
                     // playing safe here and close the engine even if the above succeeds - close can be called multiple times
                     // Also closing refreshListeners to prevent us from accumulating any more listeners
                     IOUtils.close(engine, globalCheckpointListeners, refreshListeners, pendingReplicationActions);
+
+                    // Remote Segment Store Cleanup
+                    if (deleted && isPrimaryMode() && isRemoteStoreEnabled()) {
+                        RemoteSegmentStoreDirectory remoteDirectory = getRemoteDirectory();
+                        remoteDirectory.deleteStaleSegments(0);
+                        remoteDirectory.deleteIfEmpty();
+                    }
+
+                    if (deleted && engine != null) {
+                        // Translog Clean up
+                        engine.translogManager().onDelete();
+                    }
+
                     indexShardOperationPermits.close();
                 }
             }
         }
     }
 
+    private RemoteSegmentStoreDirectory getRemoteDirectory() {
+        assert indexSettings.isRemoteStoreEnabled();
+        assert remoteStore.directory() instanceof FilterDirectory : "Store.directory is not an instance of FilterDirectory";
+        FilterDirectory remoteStoreDirectory = (FilterDirectory) remoteStore.directory();
+        FilterDirectory byteSizeCachingStoreDirectory = (FilterDirectory) remoteStoreDirectory.getDelegate();
+        final Directory remoteDirectory = byteSizeCachingStoreDirectory.getDelegate();
+        return ((RemoteSegmentStoreDirectory) remoteDirectory);
+    }
+
```

Comment:
Lets not expose Directory to IndexShard, that makes the point of having a Store moot. Can we instead have this logic in Store and just do `Store#close`

Reply:
created https://github.com/opensearch-project/OpenSearch/issues/8003 so as not to expose `RemoteDirectory` in `IndexShard` . 

`Store#close` is getting called while closing the store, which happens when shard is relocated out of the node. However this is deletion of the store, hence we can't hook it there.

Reply:
Moved segment deletion to `Store#close` in latest commit.